### PR TITLE
stream parse large types files

### LINKED ISSUE #9: Error reading types file: Cannot create a string longer than 0x1fffffe8 characters
In our monorepo, we have 25k+ TS files, this leads to `types.json` weighing 700MB+ (`trace.json` is ~40MB).

It seems that `readFile` has a limit if 512 MB (0x1fffffe8 characters).

https://github.com/microsoft/typescript-analyze-trace/blob/fbca82e26714dc44ac76d667a1b67072c1c30de4/src/analyze-trace-file.ts#L286-L291

Was https://github.com/amcasey/ts-analyze-trace/issues/10 from @jakub-g

### PR DESCRIPTION
**Pull request by mikeduminy**
Fixes #9 

I ran into this issue in a monorepo with a types.json trace file of over 600MB. Here's a simple fix for it using existing dependencies. 

Please let me know if I should make any changes 🙏

### CODE REVIEW

**Comment by mikeduminy** [comment]
@microsoft-github-policy-service agree

File: src/analyze-trace-utilities.ts [hunk 1]
```diff
@@ -8,6 +8,9 @@ import normalizePositions = require("./normalize-positions");
 import simplify = require("./simplify-type");
 import { EventSpan, ParseResult } from "./parse-trace-file";
 
+// @ts-ignore - no types
+import jsonstream = require("jsonstream-next");
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:12]: Let's wait on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64292

File: src/analyze-trace-utilities.ts [hunk 2]
```diff
@@ -151,18 +154,37 @@ export function unmangleCamelCase(name: string) {
...
-            typesCache = JSON.parse(json);
-        }
-        catch (e: any) {
-            console.error(`Error reading types file: ${e.message}`);
-            typesCache = [];
-        }
+        return new Promise((resolve, reject) => {
+            try {
+                const readStream = fs.createReadStream(typesPath, { encoding: "utf-8" });
+                readStream.on('open', () => {
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:160]:
```suggestion
                readStream.on("open", () => {
```

File: src/analyze-trace-utilities.ts [hunk 3]
```diff
@@ -151,18 +154,37 @@ export function unmangleCamelCase(name: string) {
...
+                    typesCache = []
+                })
+                readStream.on('end', () => {
+                    resolve(typesCache!)
+                });
+                readStream.on('error', (e) => {
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:166]:
```suggestion
                readStream.on("error", (e) => {
```

File: src/analyze-trace-utilities.ts [hunk 4]
```diff
@@ -151,18 +154,37 @@ export function unmangleCamelCase(name: string) {
...
+                    console.error(`Error reading types file: ${e.message}`);
+                    reject()
+                })
+
+                // expects types file to be {object[]}
+                const parser = jsonstream.parse('*')
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:172]:
```suggestion
                const parser = jsonstream.parse("*")
```

File: src/analyze-trace-utilities.ts [hunk 5]
```diff
@@ -151,18 +154,37 @@ export function unmangleCamelCase(name: string) {
...
+                parser.on('data', (data: object) => {
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:173]:
```suggestion
                parser.on("data", (data: object) => {
```

File: src/analyze-trace-utilities.ts [hunk 6]
```diff
@@ -151,18 +154,37 @@ export function unmangleCamelCase(name: string) {
...
+                });
+                readStream.on("error", (e) => {
+                    console.error(`Error reading types file: ${e.message}`);
+                    reject()
+                })
+
+                // expects types file to be {object[]}
+                const parser = jsonstream.parse("*")
+                parser.on("data", (data: object) => {
+                    (typesCache as any[]).push(data);
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:174]:
```suggestion
                    typesCache!.push(data);
```

File: src/analyze-trace-utilities.ts [hunk 7]
```diff
@@ -151,18 +154,37 @@ export function unmangleCamelCase(name: string) {
...
+                });
+
+                readStream.pipe(parser)
+            }
+            catch (e: any) {
+                console.error(`Error reading types file: ${e.message}`);
+                typesCache = [];
+                reject()
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:182]: Do we need the `try`/`catch` anymore? Won't this be handled in the `error` case?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: src/analyze-trace-utilities.ts (hunk 6 above)
Comment (mikeduminy) [reply src/analyze-trace-utilities.ts:174]: ```Property 'push' does not exist on type 'readonly any[]'.```

File: src/analyze-trace-utilities.ts (hunk 1 above)
Comment (mikeduminy) [reply src/analyze-trace-utilities.ts:12]: Totally possible to wait, but it's also just an additional usage of this existing library

File: src/analyze-trace-utilities.ts (hunk 7 above)
Comment (mikeduminy) [reply src/analyze-trace-utilities.ts]: You're right - I just removed the `try`/`catch`.

Also aligned the error handling with the original code where just an empty array is returned (the error is not propagated).

File: src/analyze-trace-utilities.ts [hunk 8]
```diff
@@ -151,18 +154,32 @@ export function unmangleCamelCase(name: string) {
...
+
+            function onError(e: Error) {
+                console.error(`Error reading types file: ${e.message}`);
+                resolve(typesCache!)
+            }
+        })
     }
 
-    return typesCache!;
+    return Promise.resolve(typesCache);
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:182]: This is an async function - I think you can just return `typesCache` here, right?

File: src/analyze-trace-utilities.ts (hunk 6 above)
Comment (DanielRosenwasser) [reply src/analyze-trace-utilities.ts]: It's probably fine as-is. My choice would actually be to create a local that's just an `any[]` and then assign it to `typesCache` - but it's no big deal.

File: src/analyze-trace-utilities.ts [hunk 9]
```diff
@@ -151,18 +154,32 @@ export function unmangleCamelCase(name: string) {
...
+            });
+            parser.on("error", onError)
+
+            readStream.pipe(parser)
+
+            function onError(e: Error) {
+                console.error(`Error reading types file: ${e.message}`);
+                resolve(typesCache!)
+            }
+        })
```
Comment (DanielRosenwasser) [inline src/analyze-trace-utilities.ts:179]:
```suggestion
            typesCache = [];

            const readStream = fs.createReadStream(typesPath, { encoding: "utf-8" });
            readStream.on("end", () => {
                resolve(typesCache!);
            });
            readStream.on("error", onError);

            // expects types file to be {object[]}
            const parser = jsonstream.parse("*");
            parser.on("data", (data: object) => {
                (typesCache as any[]).push(data);
            });
            parser.on("error", onError);

            readStream.pipe(parser);

            function onError(e: Error) {
                console.error(`Error reading types file: ${e.message}`);
                resolve(typesCache!);
            }
        });
```

File: src/analyze-trace-utilities.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/analyze-trace-utilities.ts:12]: Hah - I totally missed that this was already used by our package.

File: src/analyze-trace-utilities.ts (hunk 9 above)
Comment (mikeduminy) [reply src/analyze-trace-utilities.ts]: Sorry for missing all of these style issues 🙏

File: src/analyze-trace-utilities.ts (hunk 8 above)
Comment (mikeduminy) [reply src/analyze-trace-utilities.ts]: fixed

File: src/analyze-trace-utilities.ts (hunk 6 above)
Comment (mikeduminy) [reply src/analyze-trace-utilities.ts]: Okay keeping as-is then

File: src/analyze-trace-utilities.ts (hunk 1 above)
Comment (mikeduminy) [reply src/analyze-trace-utilities.ts:12]:
```
src/simplify-types-file.ts:63:19 - error TS2339: Property '_flush' does not exist on type 'ReadWriteStream'.

63         transform._flush = newFlush;
```

Comment (mikeduminy) [reply src/analyze-trace-utilities.ts:12]: Looks like the return type of `jsonstream.parse()` is the return of `@types/through2`

Comment (mikeduminy) [reply src/analyze-trace-utilities.ts:12]: How would you feel about adopting the types in a separate PR?

Comment (jakebailey) [reply src/analyze-trace-utilities.ts:12]: I'd sooner see us write `(transform as any)._flush` (to make it abundantly clear we're doing something unsafe) for that code than not use the types at all.

Comment (DanielRosenwasser) [reply src/analyze-trace-utilities.ts:12]: Before I merge, I'll need to try it out locally - but by that point, https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64383 will probably be merged anyway.

**Review by DanielRosenwasser — APPROVED** [review]

**Review by jakebailey — APPROVED** [review]

**Comment by mikeduminy** [comment]
Thanks for adding the types and getting this merged 💪

**Pull request merged**